src/main/kotlin/com/terraformation/backend/nursery/model/WithdrawalModel.kt

```diff
@@ -0,0 +1,81 @@
+package com.terraformation.backend.nursery.model
+
+import com.terraformation.backend.db.default_schema.FacilityId
+import com.terraformation.backend.db.nursery.BatchId
+import com.terraformation.backend.db.nursery.WithdrawalId
+import com.terraformation.backend.db.nursery.WithdrawalPurpose
+import com.terraformation.backend.db.nursery.tables.pojos.BatchWithdrawalsRow
+import com.terraformation.backend.db.nursery.tables.pojos.WithdrawalsRow
+import java.time.LocalDate
+
+/**
+ * The number of seedlings of each type that came from a single batch as part of a withdrawal. There
+ * can be more than one of these per withdrawal.
+ */
+data class BatchWithdrawalModel(
+    val batchId: BatchId,
+    val destinationBatchId: BatchId? = null,
+    val germinatingQuantityWithdrawn: Int,
+    val notReadyQuantityWithdrawn: Int,
+    val readyQuantityWithdrawn: Int,
+) {
+  init {
+    if (germinatingQuantityWithdrawn < 0 ||
+        notReadyQuantityWithdrawn < 0 ||
+        readyQuantityWithdrawn < 0) {
+      throw IllegalArgumentException("Withdrawal quantities may not be negative")
+    }
+  }
+}
+
+/**
+ * Information about a withdrawal as a whole.
+ *
+ * @param ID This class can represent a withdrawal we loaded from the database (in which case [ID]
+ * will be non-nullable) or a new withdrawal we want to insert (in which case [ID] will be a
+ * null-only type). Use [ExistingWithdrawalModel] and [NewWithdrawalModel] rather than specifying a
+ * type parameter at usage sites.
+ */
+data class WithdrawalModel<ID : WithdrawalId?>(
+    val batchWithdrawals: List<BatchWithdrawalModel>,
+    val destination: String? = null,
+    val destinationFacilityId: FacilityId? = null,
+    val facilityId: FacilityId,
+    val id: ID,
+    val purpose: WithdrawalPurpose,
+    val reason: String? = null,
+    val withdrawnDate: LocalDate,
+) {
+  init {
+    if (batchWithdrawals.isEmpty()) {
+      throw IllegalArgumentException("Withdrawals must come from at least one batch")
+    }
+  }
+}
+
+/** A withdrawal that hasn't been inserted into the database yet. */
+typealias NewWithdrawalModel = WithdrawalModel<Nothing?>
+
+/** A withdrawal that was loaded from the database. */
+typealias ExistingWithdrawalModel = WithdrawalModel<WithdrawalId>
```

Comment:
This is kind of an experiment I want to try out, a way to deal with the fact that you want the `id` field on this kind of model object to be non-nullable if the object was pulled from the database, but you want to reuse the same model class to pass in new data to be inserted (in which case there's no ID).

An alternative approach (other than doing `val id: WithdrawalId?` as we do in, e.g., `AccessionModel`) is to declare `NewWithdrawalModel` and `ExistingWithdrawalModel` as entirely separate classes, but then that becomes a maintenance headache as the data model evolves over time; the two could gradually fall out of sync.

Reply:
In case you aren't familiar with it, Kotlin's `Nothing` class is kind of an odd duck: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-nothing.html

Reply:
Oh interesting, so in the case of `WithdrawalModel<Nothing?>` it's impossible to instantiate the `Nothing` type, so it can only ever be null?

Reply:
That's the idea, yep.